Add support to allow lookalike characters on sanitization

charliewolfe opened the pull request:
#### 313974e6d6608538daaf50e5f52e29b55f3c7dc7
<pre>
Add support to allow lookalike characters on sanitization
<a href="https://bugs.webkit.org/show_bug.cgi?id=250992">https://bugs.webkit.org/show_bug.cgi?id=250992</a>
rdar://103329323

Reviewed by Wenson Hsieh.

Add the functionality to allow certain lookalike characters on sanitization when
accessed from bindings. See below for more details.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::adjustedURL const):
(WebCore::Document::urlForBindings const):

Refactor returning the adjusted URL such that we request the sanitized URL
with the allowed lookalike characters.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::allowedLookalikeCharacters const):

Add a client function to sanitize a given URL.

* Source/WebCore/page/LookalikeCharactersSanitizationData.h: Copied from Source/WebKit/Platform/cocoa/NetworkConnectionIntegrityHelpers.h.
(WebCore::LookalikeCharactersSanitizationData::LookalikeCharactersSanitizationData):
(WebCore::LookalikeCharactersSanitizationData::lookalikeCharacters):
(WebCore::LookalikeCharactersSanitizationData::encode const):
(WebCore::LookalikeCharactersSanitizationData::decode):

Add a struct to hold data used for domain scoped lookalike character sanitization.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::allowedLookalikeCharacters const):
* Source/WebCore/page/Page.h:

Add helper functions on Page to sanitize a given URL.

* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:

Add the allowed lookalike character strings to the web page creation parameters. This
is used to retrieve the cached allowed lookalike characters in the UI process.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::initializeWebPage):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::createNewPage):
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::cachedAllowedLookalikeStrings):
(WebKit::WebPageProxy::updateAllowedLookalikeCharacterStringsIfNeeded):
* Source/WebKit/UIProcess/WebPageProxy.h:

Add functionality to update the allowed lookalike characters strings in `didCommitLoadForFrame`.
This previously existed for `cachedLookalikeStrings`, but was changed because the lookalike strings
were populated too late to apply for this first navigation within a web view. Since `allowedLookalikeStrings`
only needs to be populated after a navigation has already occured, we can have the update function in
`didCommitLoadForFrame`.

* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::allowedLookalikeCharacters const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:

* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::allowedLookalikeCharacters):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::setAllowedLookalikeCharacterStrings):

Populate the allowed lookalike character strings hashmap as domain -&gt; lookalikeCharacters.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::allowedLookalikeCharacters):
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/259541@main">https://commits.webkit.org/259541@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ce863d18a04edfe2d8dfbb956cc355210f665717

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105045 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14124 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37935 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114308 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174491 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108948 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15262 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5046 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97364 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113321 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110801 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11798 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94798 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39305 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93666 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26425 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80970 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7462 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27784 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7556 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4367 "Found 1 new test failure: fast/images/avif-heif-container-as-image.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13611 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47336 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6567 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9345 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/NetworkProcess/NetworkProcess.h:
```diff
@@ -495,6 +499,10 @@ class NetworkProcess : public AuxiliaryProcess, private DownloadManager::Client,
 
+#if ENABLE(NETWORK_CONNECTION_INTEGRITY)
+    void requestAllowedLookalikeCharacterStrings(CompletionHandler<void(const Vector<WebCore::LookalikeCharactersSanitizationData>&)>&&);
+#endif
```
Could we move this to the UI process? (Or is this only temporarily being done in the network process, due to the entitlement not being available for all WebKit apps?)

charliewolfe replied:
Yes. As discussed offline, we don't need to go through the network process here. Will change this.

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5235,6 +5236,9 @@ void WebPageProxy::didCommitLoadForFrame(FrameIdentifier frameID, FrameInfoData&
         closeSharedPreviewPanelIfNecessary();
+#endif
+#if ENABLE(NETWORK_CONNECTION_INTEGRITY)
+    updateAllowedLookalikeCharacterStringsIfNeeded();
```
Nit - indentation looks like it could use 4 more spaces here.

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8825,6 +8830,9 @@ WebPageCreationParameters WebPageProxy::creationParameters(WebProcessProxy& proc
         parameters.lookalikeCharacterStrings = cachedLookalikeStrings();
+
+    if (sessionID().isEphemeral())
+        parameters.allowedLookalikeCharacterStrings = cachedAllowedLookalikeStrings();
```
Can we unconditionally set this? I don't think we want to strictly tie this to ephemeral sessions, but it's not really possible to consult the relevant network connection integrity policy bits here.

charliewolfe committed:
Add support to allow lookalike characters on sanitization
https://bugs.webkit.org/show_bug.cgi?id=250992
rdar://103329323

Reviewed by Wenson Hsieh.

Add the functionality to allow certain lookalike characters on sanitization when
accessed from bindings. See below for more details.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::adjustedURL const):
(WebCore::Document::urlForBindings const):

Refactor returning the adjusted URL such that we request the sanitized URL
with the allowed lookalike characters.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::allowedLookalikeCharacters const):

Add a client function to sanitize a given URL.

* Source/WebCore/page/LookalikeCharactersSanitizationData.h: Copied from Source/WebKit/Platform/cocoa/NetworkConnectionIntegrityHelpers.h.
(WebCore::LookalikeCharactersSanitizationData::LookalikeCharactersSanitizationData):
(WebCore::LookalikeCharactersSanitizationData::lookalikeCharacters):
(WebCore::LookalikeCharactersSanitizationData::encode const):
(WebCore::LookalikeCharactersSanitizationData::decode):

Add a struct to hold data used for domain scoped lookalike character sanitization.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::allowedLookalikeCharacters const):
* Source/WebCore/page/Page.h:

Add helper functions on Page to sanitize a given URL.

* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:

Add the allowed lookalike character strings to the web page creation parameters. This
is used to retrieve the cached allowed lookalike characters in the UI process.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::initializeWebPage):
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::createNewPage):
(WebKit::WebPageProxy::creationParameters):
(WebKit::WebPageProxy::cachedAllowedLookalikeStrings):
(WebKit::WebPageProxy::updateAllowedLookalikeCharacterStringsIfNeeded):
* Source/WebKit/UIProcess/WebPageProxy.h:

Add functionality to update the allowed lookalike characters strings in `didCommitLoadForFrame`.
This previously existed for `cachedLookalikeStrings`, but was changed because the lookalike strings
were populated too late to apply for this first navigation within a web view. Since `allowedLookalikeStrings`
only needs to be populated after a navigation has already occured, we can have the update function in
`didCommitLoadForFrame`.

* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::allowedLookalikeCharacters const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:

* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::allowedLookalikeCharacters):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::setAllowedLookalikeCharacterStrings):

Populate the allowed lookalike character strings hashmap as domain -> lookalikeCharacters.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::allowedLookalikeCharacters):
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/259541@main
